text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,409 @@
+type = "regex"
+```
+
+A buildpack is encouraged to be as permissive as possible. This ensures that a
+buildpack author won’t have to frequently update this metadata. This should be
+balanced by buildpack authors to provide compatibility guarantees with the
+tools required to run the buildpack and for the software it is installing to
+run. Generally, we believe that most buildpacks will be compatible with the
+major versions of software they presently support and that new major versions
+of dependencies should be tested and validations should be expanded after
+testing.
+
+## Metadata Distribution
+There is no prescribed method for distributing metadata. It could be done in a
+variety of ways, including HTTPS/SFTP distributed archives, `rsync` of remote
+directories, or even distributed as an image through an image registry.
+
+For the Paketo project, this proposal suggests distributing metadata through
+images in an image registry. This allows the project to use the existing image
+registry to distribute the metadata. Image registries also have inherent
+properties that help with security, like an image cannot be modified without
+creating a new hash for the image and images can be signed (signing is out of
+scope for this proposal). In addition, images are easily versioned such that
+users can hold back updates to dependencies if desired and are easily cached.
+
+The contents of the image will contain the directory structure defined in
+[Metadata Format](#metadata-format). There should not be a top-level directory
+added, so the root of the image should contain all of the directories created
+as top-level organization names, like `com` or `org`. All of the metadata is to
+be included in a single layer. Updates to metadata will require downloading the
+entire layer again, however, it is a single layer and the size is expected to
+be small so this should be very fast.
+
+Further, this proposal suggests subdividing metadata images by project
+sub-team. Each sub-team will be given a unique reverse domain name like
+`io.paketo.java` and `io.paketo.utilities`. In this way, the project’s metadata
+can be easily combined without having conflicts.
+
+This allows users to pick and choose the metadata that’s relevant to their
+needs.
+
+## Buildpack Dependency Metadata Interface
+The interface between dependency metadata and a buildpack is simple. A single
+directory of metadata will be presented to the buildpack. It will be presented
+at the location specified by `BP_DEPENDENCY_METADATA`, which defaults to
+`/platform/deps/metadata` (the intent is to pilot and try out this or possibly
+other locations, eventually proposing an RFC with Cloud-Native buildpacks to
+standardize the location).
+
+Buildpacks do not care if there are multiple sources of metadata information,
+however, this needs to be merged and presented to the buildpack as one single
+directory. How that information is merged is outside of the scope of this
```

Comment:
I kind of feel like knowing which metadata sources would take precedence in a merge situation would be helpful for when we implement this. I'm not sure if I understand why that's out of scope

Reply:
It's really just to simplify things. By the time the buildpacks run, there should only be one directory structure. Internally, all a buildpack will see is the directory hierarchy for the metadata. No sources, no merging, no complication.

If we talk about Paketo, we'll need to distribute metadata in some way. A buildpack to distribute the metadata has been discussed, so if we did that then the buildpack would specify things like where the metadata comes from, if there are multiple sources, how those are merged, etc... That is out of scope here though, but very likely something we'll want to discuss in another RFC. Because of the scope of what's being proposed, this RFC is leaving some stuff out and I agree that we'll need more RFCs to clarify some points. We just wanted to break things down in to more approachable chunks to help move this forward.

Reply:
Alright, I think punting this to a future RFC is fair. Similarly to a point that was made above - it would be nice to align on these things so that library implementations end up looking more similar, but you're right it doesn't necessarily need to happen here.